File: week8/한지희/오픈채팅방.kt
class OpenChatRoom {
    fun solution(record: Array<String>): Array<String> {
        val answer = mutableListOf<String>()
        val userNicknameHashMap = mutableMapOf<String, String>()
        val tempLog = mutableListOf<Pair<String, Boolean>>()
        record.forEach { log ->
            val logElement = log.split(" ")
            val commend = logElement[0]
            val userId = logElement[1]

            when(commend) {
                "Enter" -> {
                    userNicknameHashMap[userId] = logElement[2]
                    tempLog.add(Pair(userId, true))
                }
                "Leave" -> {
                    tempLog.add(Pair(userId, false))
                }
                "Change" -> {
                    userNicknameHashMap[userId] = logElement[2]
                }
            }
        }

        tempLog.forEach { idStatePair ->
            if (idStatePair.second) answer.add("${userNicknameHashMap[idStatePair.first]}님이 들어왔습니다.")
            else answer.add("${userNicknameHashMap[idStatePair.first]}님이 나갔습니다.")
        }

        return answer.toTypedArray()
    }
}

File: week8/윤여준/오픈채팅방.cpp
#include<bits/stdc++.h>
using namespace std;

map<string, string> m;

vector<string> split(string str) {
    vector<string> ret;
    string temp = "";
    for(char c: str) {
        if(c == ' ') {
            ret.push_back(temp);
            temp = "";
        }
        else temp += c;
    }
    ret.push_back(temp);
    
    return ret;
}

vector<string> solution(vector<string> record) {
    vector<string> answer;
    
    for(auto str: record) {
        vector<string> cmd = split(str);
        if(cmd[0] != "Leave")
            m[cmd[1]] = cmd[2];
    }
    
    for(auto str: record) {
        vector<string> cmd = split(str);
        if(cmd[0] == "Enter")
            answer.push_back(m[cmd[1]] + "님이 들어왔습니다.");
        else if(cmd[0] == "Leave")
            answer.push_back(m[cmd[1]] + "님이 나갔습니다.");
    }
    
    return answer;
}